### PR TITLE
feat: db state sync by merk chunking

### PR DESCRIPTION
**Pull request by ogabrielides**
## Issue being fixed or feature implemented
Implementation of grovedb state sync.

## What was done?
Work based on top of PR #267.

## How Has This Been Tested?
Introduced new tutorial `replication`

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: grovedb/src/replication.rs [hunk 1]
```diff
@@ -1,990 +1,430 @@
...
-
-/// Structure to drive GroveDb restore process.
-pub struct Restorer<'db> {
-    current_merk_restorer: Option<MerkRestorer<'db>>,
-    current_merk_chunk_index: usize,
-    current_merk_path: Path,
-    queue: VecDeque<(Path, Vec<u8>, Hash, TreeFeatureType)>,
-    grove_db: &'db GroveDb,
-    tx: &'db Transaction<'db>,
+// Converts a path into a human-readable string (for debuting)
```
Comment (QuantumExplorer) [inline grovedb/src/replication.rs:74]: debugging

File: grovedb/src/replication.rs [hunk 2]
```diff
@@ -1,990 +1,430 @@
...
+pub fn util_path_to_string(path: &[Vec<u8>]) -> Vec<String> {
+    let mut subtree_path_str: Vec<String> = vec![];
+    for subtree in path {
+        let string = std::str::from_utf8(subtree).unwrap();
```
Comment (QuantumExplorer) [inline grovedb/src/replication.rs:78]: expect

File: grovedb/src/replication.rs [hunk 3]
```diff
@@ -1,990 +1,430 @@
...
+        subtree_path_str.push(string.parse().unwrap());
```
Comment (QuantumExplorer) [inline grovedb/src/replication.rs:79]: expect

File: grovedb/src/replication.rs [hunk 4]
```diff
@@ -1,990 +1,430 @@
...
-/// successful result.
-#[derive(Debug)]
-pub enum RestorerResponse {
-    AwaitNextChunk { path: Vec<Vec<u8>>, index: usize },
-    Ready,
-}
+// Splits the given global chunk id into [SUBTREE_PREFIX:CHUNK_ID]
+pub fn util_split_global_chunk_id(
+    global_chunk_id: &[u8],
+) -> Result<(crate::SubtreePrefix, String), Error> {
```
Comment (QuantumExplorer) [inline grovedb/src/replication.rs:87]: why is this a string and not a Vec<u8>?

File: grovedb/src/replication.rs [hunk 5]
```diff
@@ -1,990 +1,430 @@
...
+
+            let current_path = SubtreePath::from(path);
+
+            match (current_path.derive_parent(), subtree.last()) {
+                (Some((parent_path, _)), Some(parent_key)) => match tx {
+                    None => {
+                        let parent_merk = self
+                            .open_non_transactional_merk_at_path(parent_path, None)
+                            .value?;
+                        if let Ok((Some((elem_value, elem_value_hash)))) = parent_merk
```
Comment (QuantumExplorer) [inline grovedb/src/replication.rs:147]: clippy

File: grovedb/src/replication.rs [hunk 6]
```diff
@@ -1,990 +1,430 @@
...
-                            value_bytes.to_owned(),
-                            *value_hash,
-                            *feature_type,
-                        ));
                     }
+                    Some(t) => {
+                        let parent_merk = self
+                            .open_transactional_merk_at_path(parent_path, t, None)
+                            .value?;
+                        if let Ok((Some((elem_value, elem_value_hash)))) = parent_merk
```
Comment (QuantumExplorer) [inline grovedb/src/replication.rs:166]: clippy

File: grovedb/src/replication.rs [hunk 7]
```diff
@@ -1,990 +1,430 @@
...
+    // LoadSnapshotChunk method is called) Params:
+    // global_chunk_id: Global chunk id in the following format:
+    // [SUBTREE_PREFIX:CHUNK_ID] SUBTREE_PREFIX: 32 bytes (mandatory) (All zeros
+    // = Root subtree) CHUNK_ID: 0.. bytes (optional) Traversal instructions to
+    // the root of the given chunk. Traversal instructions are "1" for left, and
+    // "0" for right. TODO: Compact CHUNK_ID into bitset for size optimization
+    // as a subtree can be big hence traversal instructions for the deepest chunks
+    // tx: Transaction. Function returns the data by opening merks at given tx.
+    // Returns the Chunk proof operators for the requested chunk
+    pub fn fetch_chunk<'db>(
```
Comment (QuantumExplorer) [inline grovedb/src/replication.rs:207]: Elide 'db

File: grovedb/src/replication.rs [hunk 8]
```diff
@@ -1,990 +1,430 @@
...
+                let subtree_path: Vec<&[u8]> = subtree.iter().map(|vec| vec.as_slice()).collect();
+                let path: &[&[u8]] = &subtree_path;
 
-impl<'db> BufferedRestorer<'db> {
-    /// New
-    pub fn new(restorer: Restorer<'db>) -> Self {
-        BufferedRestorer { restorer }
-    }
+                let merk = self
+                    .open_non_transactional_merk_at_path(path.into(), None)
```
Comment (QuantumExplorer) [inline grovedb/src/replication.rs:234]: we are opening a non transactional, even if a transaction is set.

File: grovedb/src/replication.rs [hunk 9]
```diff
@@ -1,990 +1,430 @@
...
-                    .expect("cannot get next chunk");
-                match restorer.process_chunk(chunk).expect("cannot process chunk") {
-                    RestorerResponse::Ready => break,
-                    RestorerResponse::AwaitNextChunk { path, index } => {
-                        next_chunk = (path, index);
+                let chunk_producer_res = ChunkProducer::new(&merk);
+                match chunk_producer_res {
+                    Ok(mut chunk_producer) => {
+                        let chunk_res = chunk_producer
+                            .chunk(String::from_utf8(chunk_id.to_vec()).unwrap().as_str());
```
Comment (QuantumExplorer) [inline grovedb/src/replication.rs:245]: unwrap in production code.

File: merk/src/merk/chunks.rs [hunk 10]
```diff
@@ -26,479 +26,1041 @@
...
-                ));
+    /// Packs as many chunks as it can from a starting chunk index, into a
+    /// vector. Stops when we have exhausted all chunks or we have reached
+    /// some limit.
+    fn subtree_multi_chunk_with_limit(
+        &mut self,
+        index: usize,
+        limit: Option<usize>,
+    ) -> Result<SubtreeChunk, Error> {
+        let mut chunk_byte_length = 0;
```
Comment (QuantumExplorer) [inline merk/src/merk/chunks.rs:285]: value assigned to `chunk_byte_length` is never read

File: merk/src/merk/restore.rs [hunk 11]
```diff
@@ -29,263 +29,284 @@
...
+    /// Initializes a new chunk restorer with the expected root hash for the
+    /// first chunk
     pub fn new(
         merk: Merk<S>,
-        combining_value: Option<Vec<u8>>,
         expected_root_hash: CryptoHash,
+        parent_key_value_hash: Option<CryptoHash>,
     ) -> Self {
+        let mut chunk_id_to_root_hash = BTreeMap::new();
+        chunk_id_to_root_hash.insert(traversal_instruction_as_string(&vec![]), expected_root_hash);
```
Comment (QuantumExplorer) [inline merk/src/merk/restore.rs:75]: &[]

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: grovedb/src/replication.rs (hunk 4 above)
Comment (ogabrielides) [reply grovedb/src/replication.rs:92]: @QuantumExplorer Existing work on merk chunking makes use of string for chunk identification.
This PR, uses Vec<u8> for chunk identification in all APIs for ABCI, but continue to use internally string in merk chunking.
I agree that this should be changed to Vec<u8> everywhere but I propose to make this change on a separate PR.

File: merk/src/proofs/tree.rs [hunk 12]
```diff
@@ -257,58 +350,30 @@ impl<'a> LayerIter<'a> {
...
+                    self.stack.push((&right_child.tree, item_depth + 1))
+                }
+                if let Some(left_child) = item.child(true) {
+                    self.stack.push((&left_child.tree, item_depth + 1))
                 }
-
-                return item;
             } else {
-                popped = self.stack.pop();
+                return Some(item);
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply merk/src/proofs/tree.rs:372]: No need to do this now.

File: grovedb/src/replication.rs (hunk 4 above)
Comment (QuantumExplorer) [reply grovedb/src/replication.rs:92]: Okay in a future PR then

**Review by QuantumExplorer — APPROVED** [review]

**Pull request merged**